libnetwork/osl/namespace_linux.go

```diff
@@ -555,32 +555,25 @@ func (n *Namespace) Restore(interfaces map[Iface][]IfaceOption, routes []*types.
 	return nil
 }
 
-// Checks whether IPv6 needs to be enabled/disabled on the loopback interface
-func (n *Namespace) checkLoV6() {
-	var (
-		enable = false
-		action = "disable"
-	)
-
-	n.mu.Lock()
-	for _, iface := range n.iFaces {
-		if iface.AddressIPv6() != nil {
-			enable = true
-			action = "enable"
-			break
-		}
-	}
-	n.mu.Unlock()
-
-	if n.loV6Enabled == enable {
-		return
+// Enable IPv6 on the loopback interface, if possible.
+func (n *Namespace) enableLoV6() {
+	if err := setIPv6(n.path, "lo", true); err != nil {
+		log.G(context.TODO()).Warnf("Failed to enable IPv6 on loopback interface on network namespace %q: %v", n.path, err)
 	}
+}
```

Comment:
Have you seen what `setIPv6()` does? It looks like all the `enableLoV6` function does is force the sysctl `net.ipv6.conf.lo.disable_ipv6=0` in the sandbox, overriding the user's wishes to `--sysctl net.ipv6.conf.lo.disable_ipv6=1`. Why does this function exist?

Reply:
Yes, I have seen what `setIPv6()` does - and no, it doesn't override the user's wishes.

`enableLoV6()` exists to make sure IPv6 is enabled in the container's namespace, if it can be enabled - as the comment you've highlighted explains. It is called during the SetKey(). The user's `--sysctl` is applied later.

After the user's sysctl has been applied, after the SetKey() round of initialisation has finished and before the container task is started, `IPv6LoEnabled()` is used to check the final state. Its comment says:

```
// IPv6LoEnabled checks whether the loopback interface has an IPv6 address ('::1'
// is assigned by the kernel if IPv6 is enabled). Namespace.enableLoV6() will try
// to enable it - but it will be disabled if, for example, IPv6 is not enabled in
// the kernel, or the container was started with "--sysctl
// net.ipv6.conf.all.disable_ipv6=1".
```

The new integration test verifies/demonstrates that the command line `--sysctl` does take effect.

You'll also see in the diff that `enableLoV6()` replaces `checkLoV6()` - which called `setIPv6()` to set `disable_ipv6` on `lo` depending on whether the network being connected had IPv6 enabled ... that's the behaviour we want to get away from - the container's IPv6-ness shouldn't depend on what it's connected to. So, this change makes the same `disable_ipv6` update as before, but now it's unconditionally set to `0`.

Reply:
It's a real pain that [the `create` OCI lifecycle hooks are under-specified.](https://github.com/opencontainers/runtime-spec/blob/main/config.md#createRuntime-hooks:~:text=The%20definition%20of%20createRuntime,performed%20by%20the%20runtime.) It gives runC free reign to make the hooks useless for many purposes. You're right; runC applies the sysctls in the OCI spec after the `prestart` OCI hook (read: `SetKey()`) completes.

That being said, [the initial value of `net.ipv6.conf.<iface>.disable_ipv6` in a freshly-created network namespace is `0` by default.](https://www.kernel.org/doc/html/latest/networking/ip-sysctl.html#:~:text=Default%3A%2016-,disable_ipv6%20%2D%20BOOLEAN,Default%3A%20FALSE%20(enable%20IPv6%20operation),-When%20this%20value) However, [there is a sysctl `devconf_inherit_init_net` to control this behaviour.](https://www.kernel.org/doc/html/v6.6/admin-guide/sysctl/net.html#devconf-inherit-init-net) If the sysctl is left at the default, `setIPv6()` is a no-op. And if the sysctl is modified, `setIPv6()` may override the user's wishes.

Further complicating matters is the `docker run --net=container:<ref>` flag to join a new container to another container's netns (as made famous by k8s Pods). `setIPv6()` called when adding a container to an existing pod may force-enable IPv6 in the netns after the user explicitly disabled it with `--sysctl` on the first container.

tl;dr `setIPv6()` is either a no-op, or forces IPv6 enabled when the user doesn't expect it.

Reply:
Yes, if the default is IPv6-enabled , I can remove it. (Although it doesn't run for the container network / '--net=container' case, as discussed above.)

So ...

Checking for the '::1' is enough to detect a kernel that doesn't have IPv6, or any means the user has of disabling IPv6 under the daemon's feet.

If we parse the '--sysctl' option, and treat it as a "no IPv6" flag, the IPv6 config can all happen in SetKey. The '::1' still being present because of the early restart hook wouldn't matter.

(The libnetwork logic can then be simplified a bit, because the SetKey part will always run.)

Reply:
I've done that ... removing several `setIPv6()` calls - including this one, and calls to `disable_ipv6` for `all` as a "starting point" when creating containers.